### PR TITLE
page_service: getpage batching: refactor & minor fixes

### PR DESCRIPTION
**Pull request by problame**
This PR refactors the page_service server-side batching code that was recently added in
https://github.com/neondatabase/neon/issues/9377.

Changes:

- Store carried-over batching state in a local variable inside `handle_pagestream`, instead of in the `PageServerHandler`. This adds robustness because it systematically avoids a source of bugs.
- Instead of `next_batch`, call it `carry`.
- When starting a new batch read, take into account the time carried over from the previous call.
- Build the batch inside the `&mut Option<Carry>` of in a local variable.
  - Before, we would have to make sure that we move the `batch` back into the `self.next_batch` whenever we bail.

### CODE REVIEW

**Review by VladLazar — APPROVED** [review]
Checked it out locally and stepped through it a number of times. Looks correct, but have a look at the comments.

File: pageserver/src/page_service.rs [hunk 1]
```diff
@@ -611,55 +612,99 @@ impl PageServerHandler {
         )
     }
 
+    #[instrument(skip_all, level = tracing::Level::TRACE)]
     async fn read_batch_from_connection<IO>(
```
[the remark this answers is not in the document]
Comment (problame) [reply pageserver/src/page_service.rs:616]: https://github.com/neondatabase/neon/issues/9834

File: pageserver/src/page_service.rs [hunk 2]
```diff
@@ -611,55 +612,99 @@ impl PageServerHandler {
...
+            // Create a future that will become ready when we need to stop batching.
+            use futures::future::Either;
+            let batching_deadline = match (
+                &*maybe_carry as &Option<Carry>,
+                &mut batching_deadline_storage,
+            ) {
+                (None, None) => Either::Left(futures::future::pending()), // there's no deadline before we have something batched
+                (None, Some(_)) => unreachable!(),
+                (Some(_), Some(fut)) => Either::Right(fut), // below arm already ran
+                (Some(carry), None) => {
```
[the remark this answers is not in the document]
Comment (problame) [reply pageserver/src/page_service.rs:641]: Good find, yeah.

I think that's a pre-existing issue before this patch, right?

I'll create a follow-up ticket and get back to it after I have found a solution to enforce small batching timeouts (currently being explored in https://github.com/neondatabase/neon/pull/9822 )

[the remark this answers is not in the document]
Comment (problame) [reply pageserver/src/page_service.rs:641]: = https://github.com/neondatabase/neon/issues/9835

[the remark this answers is not in the document]
Comment (VladLazar) [reply pageserver/src/page_service.rs:641]: > I think that's a pre-existing issue before this patch, right?

Yeah, it is.

**Comment by problame** [comment]
The changes in this PR will be largely replaced by

- https://github.com/neondatabase/neon/pull/9851

which is currently stacked on top.

Closing this to avoid one merge-rebase-CI roudtrip.

**Pull request closed without merging**